Remove m prefix requirement

josibake opened the pull request:
`m` in BIP0032 is a variable, not a constant. Requiring it as a constant here is confusing and can lead to erroneous conclusions if using this library as a means of understanding BIP0032.
 
Fixes #2449

tcharding commented:
Any reason not to make the "m/" optional, would save changing all the unit tests and "m/0/1"  reads pretty clearly?

Kixunil commented:
@tcharding it was just wrong originally.

josibake commented on bitcoin/src/bip32.rs:
```diff
@@ -450,8 +449,11 @@ impl DerivationPath {
-        for cn in self.0.iter() {
+        let mut iter = self.0.iter();
+        if let Some(first_element) = iter.next() {
+            fmt::Display::fmt(first_element, f)?;
```
Still getting used to rust so pardon my ignorance, but this shouldn't be changing anything. Previously, the code was writing `m` directly to the formatter and then calling `fmt::Display::fmt(cn, f)?` in a loop. My understanding is `fmt::Display::fmt(cn, f)` is being called so that the `ChildNumber` types get printed with the hardened notation.

Since we no longer start the strings with `m` my change is consuming the first element from the iterator (which is a `ChildNumber`) and calling `fmt::Display::fmt` so that the first element is properly displayed with the hardened notation before proceeding with the rest of the elements.

I'm probably misunderstanding your point, but AFAICT this is keeping the behavior for displaying derivation paths the same as before.

Kixunil replied:
Formatter can have options that are passed from the format string (e.g. `{:42}` pads the content to 42 chars). By passing formatter directly you keep the format options so if the user writes `{:5}` Instead of `  1/2` your code will output `    1/    2`.

My suggestion removes formatting options so they won't work at all which I think is better than having it broken. We can improve it later.

apoelstra replied:
The old code was also wrong :).

The issue is that if you try to `format!("{:20}", path)` say, the existing code (and your modification) will apply the `:20` to each of the child numbers, instead of applying it to the whole path.

josibake replied:
Gotcha! Thanks for the explanation @Kixunil @apoelstra , will update to use @Kixunil 's suggestion.

josibake commented:
Regarding the CI failures, it looks like all of them are due to running an old version of the `ecdsa-psbt` example. Any recommendations on how to go about fixing this?

apoelstra commented:
> @tcharding it was just wrong originally.

FWIW I think the BIP is pretty confusingly written; you need to find the words "To shorten notation" in the middle of the body in order to realize that the unit tests shouldn't be taken literally? I guess a PR would be welcome.

Kixunil commented:
Behold, CI failing with a shitty error that nobody knows where it comes from - this is why I'm pushing for adding information to errors.

tcharding commented:
> a shitty error 

Its shitty yes because it can come from two code paths in the same function, but I believe our current error strategy, if applied to the `bip32` module, would fix this.

> that nobody knows where it comes from

It only comes from a single function in the whole codebase, so this is probably a bit hyperbolic :)

tcharding commented:
> Spent some time looking at the tests to see if there's an obvious way to avoid the failures with the old versions of the ecdsa-psbt example

I just remove the "m/" from ` let path = "m/84h/0h/0h".into_derivation_path()?;` and the example runs.

Kixunil commented:
> Its shitty yes because it can come from two code paths in the same function

Actually I have no idea which input it failed on

> I believe our current error strategy, if applied to the `bip32` module, would fix this.

Yes.

> It only comes from a single function in the whole codebase, so this is probably a bit hyperbolic

I couldn't find the place where `m/` is used but sure, I didn't look crazy hard.

josibake commented:
> You need to squash the commits since we require all commits to pass CI.

Ah, I see. `cargo test` was passing locally on the first commit, which is why I deferred the example changes to a separate commit, but the CI was running the examples on the first commit as well. When looking at the CI output, I saw an `--edition=2021` flag being passed and incorrectly assumed this was being used to run an old version of the example with a new version of the code, but turns out I was overthinking it :sweat_smile: 

> I just removed the "m/" from  let path = "m/84h/0h/0h".into_derivation_path()?; and the example runs.

Looks like I also missed a few "m/"! I removed the remaining "m/"'s from the ecdsa-psbt example and also from the taproot-psbt example.

Squashed per @Kixunil 's suggestion, removed the remaining "m/"'s per @tcharding . Thanks for the review!

Kixunil commented on bitcoin/src/bip32.rs:
```diff
@@ -450,10 +449,13 @@ impl DerivationPath {
+        for cn in iter {
             f.write_str("/")?;
-            fmt::Display::fmt(cn, f)?;
+            write!(f, "{}", cn)?;
```
`write!(f, "/{}", cn)?;` would've been simpler but not a big deal.

Kixunil reviewed (approved):
ACK ccbd09d5fbdb7ec26362b6b6719755e000f56e9c

apoelstra reviewed (approved):
ACK ccbd09d5fbdb7ec26362b6b6719755e000f56e9c
